### PR TITLE
silverbullet: init at 0.7.6

### PR DESCRIPTION
**Pull request by aorith**
## Description of changes

Init Silverbullet at 0.7.6
https://github.com/silverbulletmd/silverbullet

Also included a module and its tests.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [x] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Comment by aorith** [comment]
Result of `nixpkgs-review pr 291346` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>silverbullet</li>
  </ul>
</details>

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

File: pkgs/by-name/si/silverbullet/package.nix [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+  nativeBuildInputs = [ ];
+
+  buildInputs = [ ];
+
+  dontUnpack = true;
+
+  installPhase = ''
+    mkdir -p $out/{bin,lib}
+    cp $src $out/lib/silverbullet.js
+    printf '#!/usr/bin/env sh\nexec %s run -A --unstable-kv --unstable-worker-options %s "$@"\n' "${deno}/bin/deno" "$out/lib/silverbullet.js" > $out/bin/silverbullet
```
[the remark this answers is not in the document]
Comment (aorith) [reply pkgs/by-name/si/silverbullet/package.nix]: Thank you for the review.
I've externalized the script as suggested, though I'm not sure if the way I did it is what you had in mind.

File: pkgs/by-name/si/silverbullet/package.nix [hunk 2]
```diff
@@ -0,0 +1,39 @@
+{
+  lib,
+  stdenv,
+  fetchurl,
+  deno,
+}:
```
Comment (lolbinarycat) [inline pkgs/by-name/si/silverbullet/package.nix:6]:
```suggestion
{ lib
, stdenv
, fetchurl
, deno
}:
```

this is the conventional formatting for nixpkgs

File: pkgs/by-name/si/silverbullet/package.nix [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+stdenv.mkDerivation (
+  finalAttrs: {
+    pname = "silverbullet";
+    version = "0.7.3";
+
+    src = fetchurl {
+      url = "https://github.com/silverbulletmd/silverbullet/releases/download/${finalAttrs.version}/${finalAttrs.pname}.js";
```
Comment (lolbinarycat) [inline pkgs/by-name/si/silverbullet/package.nix:13]:
```suggestion
      url = "https://github.com/silverbulletmd/silverbullet/releases/download/${finalAttrs.version}/silverbullet.js";
```

interpolating `finalAttrs.pname` can cause issues when overriding the package name.

File: pkgs/by-name/si/silverbullet/package.nix [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+    dontUnpack = true;
+
+    installPhase = ''
+      mkdir -p $out/{bin,lib}
+      cp $src $out/lib/silverbullet.js
+      cp ${./silverbullet} $out/bin/silverbullet
+      substituteInPlace $out/bin/silverbullet \
+        --replace "%deno%" "${deno}/bin/deno" \
+        --replace "%silverbulletjs%" "$out/lib/silverbullet.js"
+      chmod +x $out/bin/silverbullet
```
Comment (lolbinarycat) [inline pkgs/by-name/si/silverbullet/package.nix:26]:
```suggestion
      runHook preInstall
      mkdir -p $out/{bin,lib}
      cp $src $out/lib/silverbullet.js
      cp ${./silverbullet} $out/bin/silverbullet
      substituteInPlace $out/bin/silverbullet \
        --replace "%deno%" "${deno}/bin/deno" \
        --replace "%silverbulletjs%" "$out/lib/silverbullet.js"
      chmod +x $out/bin/silverbullet
      runHook postInstall
```

you must run manually run phase hooks when overriding a phase

**Review by lolbinarycat — CHANGES_REQUESTED** [review]

File: pkgs/by-name/si/silverbullet/package.nix (hunk 2 above)
Comment (aorith) [reply pkgs/by-name/si/silverbullet/package.nix:6]: thanks, I used `nixfmt-rfc-style` too soon

**Review by lolbinarycat — APPROVED** [review]
haven't tested anything on my machine, but as long as patchShebangs is being run by the correct phases, everything should be good.

**Comment by aorith** [comment]
Thanks, I tested both in Linux and MacOs, actually as long as we have a version of Deno compatible in nixpkg it will work fine. There are binary releases (it's a work in progress) but those will be harder to port to Nix and also support less architectures than using Deno directly, it's also the same way the official docker images are using now.

File: pkgs/by-name/si/silverbullet/silverbullet [hunk 5]
```diff
@@ -0,0 +1,3 @@
+#!/usr/bin/env sh
```
Comment (AndersonTorres) [inline pkgs/by-name/si/silverbullet/silverbullet:1]: I am almost certain that this script can be generated by makeWrapper/wrapProgram.

File: pkgs/by-name/si/silverbullet/package.nix [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+
+    installPhase = ''
+      runHook preInstall
+      mkdir -p $out/{bin,lib}
+      cp $src $out/lib/silverbullet.js
+      cp ${./silverbullet} $out/bin/silverbullet
+      substituteInPlace $out/bin/silverbullet \
+        --replace "%deno%" "${deno}/bin/deno" \
+        --replace "%silverbulletjs%" "$out/lib/silverbullet.js"
+      chmod +x $out/bin/silverbullet
```
Comment (AndersonTorres) [inline pkgs/by-name/si/silverbullet/package.nix:26]: Try this instead:

```suggestion
  makeWrapper ${lib.getExe deno} $out/bin/silverbullet \
    --set DENO_NO_UPDATE_CHECK "1" \
   --append-flags "run -A --unstable-kv --unstable-worker-options ${placeholder "out"}/lib/silverbullet.js"
```

File: pkgs/by-name/si/silverbullet/package.nix [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+      runHook postInstall
+    '';
+
+    meta = with lib; {
```
Comment (AndersonTorres) [inline pkgs/by-name/si/silverbullet/package.nix:30]:
```suggestion
    meta = {
```

https://github.com/NixOS/nixpkgs/issues/208242

File: pkgs/by-name/si/silverbullet/package.nix [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+      changelog = "https://github.com/silverbulletmd/silverbullet/blob/${finalAttrs.version}/website/CHANGELOG.md";
+      description = "An open-source, self-hosted, offline-capable Personal Knowledge Management (PKM) web application";
+      homepage = "https://silverbullet.md";
+      license = licenses.mit;
+      maintainers = with lib.maintainers; [ aorith ];
```
Comment (AndersonTorres) [inline pkgs/by-name/si/silverbullet/package.nix:35]:
```suggestion
      license = lib.licenses.mit;
      maintainers = with lib.maintainers; [ aorith ];
```

**Review by AndersonTorres — CHANGES_REQUESTED** [review]

**Comment by aorith** [comment]
I have a module + tests and a version bump "ready".
I'll add that in another PR if this one gets merged.

File: pkgs/by-name/si/silverbullet/silverbullet [hunk 9]
```diff
@@ -0,0 +1 @@
...
```
Comment (AndersonTorres) [inline pkgs/by-name/si/silverbullet/silverbullet:1]: This file can be deleted, I suppose.

**Comment by AndersonTorres** [comment]
> I have a module + tests and a version bump "ready". I'll add that in another PR if this one gets merged.

Do it now

File: pkgs/by-name/si/silverbullet/silverbullet (hunk 9 above)
Comment (aorith) [reply pkgs/by-name/si/silverbullet/silverbullet]: Oops, that's right

**Comment by aorith** [comment]
Ok, so I went ahead and added the module, tests and version bump.
Hopefully it doesn't require a lot of work for the review :pray: 

The module has be basic options defined and as the application evolves I think that using `extraArgs` and `envFile` will avoid the need to update it frequently.
The test case ensures that a simple `enable = true;` works and another one with most of the options set does too.

Upstream doesn't create the space directory automatically but the module does, if it's under `/var/lib/` then systemd `StateDirectory` takes care, if not, it executes `mkdir -p`, I don't think that it would hurt to create it.

I have ran nixpkgs-review after the changes:

Result of `nixpkgs-review pr 291346` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package blacklisted:</summary>
  <ul>
    <li>nixos-install-tools</li>
  </ul>
</details>
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>silverbullet</li>
  </ul>
</details>

**Review by AndersonTorres — APPROVED** [review]

**Review by AndersonTorres — COMMENTED** [review]
Split the commit according to contributing guidelines.

More precisely:

- One commit for adding new maintainer
- One commit for adding the package
- One commit for NixOS modules
- One commit for NixOS tests

**Review by AndersonTorres — APPROVED** [review]
LGTM

**Comment by wegank** [comment]
@ofborg test silverbullet

**Pull request merged**